logs_sdk/lib/opentelemetry/sdk/logs/logger_provider.rb

```diff
@@ -0,0 +1,126 @@
+        attr_reader :resource
+
+        UNEXPECTED_ERROR_MESSAGE = 'unexpected error in ' \
+          'OpenTelemetry::SDK::Logs::LoggerProvider#%s'
+
+        private_constant :UNEXPECTED_ERROR_MESSAGE
+
+        # Returns a new LoggerProvider instance.
+        #
+        # @param [Resource] resource An optional resource to associate with
+        #   new LogRecords created by {Logger}s created by this LoggerProvider.
+        #
+        # @return [OpenTelemetry::SDK::Logs::LoggerProvider]
+        def initialize(resource: OpenTelemetry::SDK::Resources::Resource.create)
+          @log_record_processors = []
+          @mutex = Mutex.new
+          @resource = resource
+          @stopped = false
+        end
+
+        # Creates an {OpenTelemetry::SDK::Logs::Logger} instance.
+        #
+        # @param [String] name Instrumentation package name
+        # @param [String] version Optional instrumentation package version
+        #
+        # @return [OpenTelemetry::SDK::Logs::Logger]
+        def logger(name:, version: nil)
+          version ||= ''
+
+          if !name.is_a?(String) || name.empty?
+            OpenTelemetry.logger.warn('LoggerProvider#logger called with an ' \
+              "invalid name. Name provided: #{name.inspect}")
+          end
+
+          OpenTelemetry::SDK::Logs::Logger.new(name, version, self)
+        end
+
+        # Adds a new log record processor to this LoggerProvider's
+        # log_record_processors.
+        #
+        # @param [LogRecordProcessor] log_record_processor The
+        #   {LogRecordProcessor} to add to this LoggerProvider.
+        def add_log_record_processor(log_record_processor)
+          @mutex.synchronize do
+            if @stopped
+              OpenTelemetry.logger.warn('calling LoggerProvider#' \
+                'add_log_record_processor after shutdown.')
+              return
+            end
+            @log_record_processors = @log_record_processors.dup.push(log_record_processor)
+          end
+        end
+
+        # Attempts to stop all the activity for this LoggerProvider. Calls
+        # {LogRecordProcessor#shutdown} for all registered {LogRecordProcessor}s.
+        #
+        # This operation may block until all log records are processed. Must
+        # be called before turning off the main application to ensure all data
+        # are processed and exported.
+        #
+        # After this is called all newly created {LogRecord}s will be no-op.
+        #
+        # @param [Numeric] timeout An optional timeout in seconds.
+        # @return [Integer] Export::SUCCESS if no error occurred, Export::FAILURE if
+        #   a non-specific failure occurred, Export::TIMEOUT if a timeout occurred.
+        def shutdown(timeout: nil)
+          @mutex.synchronize do
+            if @stopped
+              OpenTelemetry.logger.warn('LoggerProvider#shutdown called multiple times.')
+              return OpenTelemetry::SDK::Logs::Export::FAILURE
```

Comment:
Does this need to be fully qualified?
```suggestion
              return Export::FAILURE
```